#**ASSIGNMENT(ML)-1**

SUBMITTED BY-SIMONI SINGH

IT-2

12601032025

##**Tasks**


In [34]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

#**Part A: Dataset Understanding**






Q1.
Load the dataset and display the first five records.

In [1]:
import pandas as pd

# Load the dataset
df = pd.read_csv('Dataset 2.csv')

# Display the first five records
display(df.head())

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


Q2.
Determine the number of rows and columns in the dataset.


In [3]:
# Determine the number of rows and columns
print('Shape',df.shape)

print('\ncolumns:')
print('Columns',df.columns)

row_count = len(df)
print(row_count)

Shape (750, 10)

columns:
Columns Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')
750


Q3.
Display all column names.


In [4]:
# Display all column names
print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


Q4.
Identify numerical and categorical features.


In [5]:
# Identify numerical and categorical features
numerical_cols = df.select_dtypes(include=['number']).columns.tolist()
print("Numerical features:", numerical_cols)

categorical_cols = df.select_dtypes(include=['object', 'category']).columns.tolist()
print("Categorical features:", categorical_cols)


Numerical features: ['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks', 'MonthlySpend']
Categorical features: ['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed']


Q5.
Check whether the dataset contains missing values.

In [6]:
# Check for missing values
missing_values = df.isnull().sum()
print("Missing values per column:")
display(missing_values)

Missing values per column:


,0
UserID,0
Age,0
Gender,0
SubscriptionType,0
WatchHoursPerWeek,0
DevicesUsed,0
FavoriteGenre,0
AdClicks,0
MonthlySpend,0
SubscriptionRenewed,0


#**Part B: Exploratory Data Analysis**

---




Q6.
Calculate the average age of users.


In [8]:
print("\nAverage Age:", df['Age'].mean())


Average Age: 41.824


Q7.
Determine the average watch hours per week.


In [9]:
print("Average Watch Hours Per Week:", df['WatchHoursPerWeek'].mean())

Average Watch Hours Per Week: 14.236


Q8.
Find the average monthly spending of users.


In [11]:
print("Average Monthly Spending:", df['MonthlySpend'].mean())

Average Monthly Spending: 689.9053333333334


Q9.
Count the number of users in each subscription category.


In [14]:
print("\nUsers in Each Subscription Category:")
print(df['SubscriptionType'].value_counts())


Users in Each Subscription Category:
SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


Q10.
Determine the percentage of users who renewed their subscriptions.

In [15]:
renewal_percentage = (
    (df['SubscriptionRenewed'] == 'Yes').sum()
    / len(df)) * 100
print("\nSubscription Renewal Percentage:")
print(round(renewal_percentage, 2), "%")



Subscription Renewal Percentage:
46.27 %


#**Part C: Data Preparation**

---




Q11.
Convert categorical features into numerical form.


In [16]:
le = LabelEncoder()

for col in ['Gender',
            'SubscriptionType',
            'FavoriteGenre',
            'SubscriptionRenewed']:
    df[col] = le.fit_transform(df[col])

print("\nEncoded Dataset:")
print(df.head())


Encoded Dataset:
   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


Q12.
Define the feature set (X) and target variable (y) for subscription renewal prediction.


In [17]:
X = df.drop('SubscriptionRenewed', axis=1)
y = df['SubscriptionRenewed']


Q13.
Split the dataset into training and testing sets.

In [18]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

#**Part D: Decision Tree Classification**

---




Q14.
Train a Decision Tree model to predict whether a user will renew their subscription.

Q15.
Evaluate the model using accuracy.


In [33]:
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)

dt_accuracy = accuracy_score(y_test, y_pred_dt)

print("\nDecision Tree Accuracy:")
print(round(dt_accuracy * 100, 2), "%")


Decision Tree Accuracy:
56.67 %


Q16.
Generate and interpret the confusion matrix.

In [23]:
cm = confusion_matrix(y_test, y_pred_dt)

print("\nConfusion Matrix:")
print(cm)



Confusion Matrix:
[[50 32]
 [33 35]]


#**Part E: K-Nearest Neighbors (KNN)**

---




Q17.
Train a KNN classifier with K = 5.


In [26]:
from sklearn.neighbors import KNeighborsClassifier

# Create the model with K = 5
model = KNeighborsClassifier(n_neighbors=5)

# Train the model with your data
model.fit(X_train, y_train)

# Make predictions on new data
predictions = model.predict(X_test)


### Q18. Compare the accuracy of KNN with the Decision Tree model.

In [27]:
y_pred_knn = knn.predict(X_test)

knn_accuracy = accuracy_score(y_test, y_pred_knn)

print("\nKNN Accuracy:")
print(round(knn_accuracy * 100, 2), "%")

print("\nAccuracy Comparison")
print("Decision Tree:", round(dt_accuracy * 100, 2), "%")
print("KNN:", round(knn_accuracy * 100, 2), "%")


KNN Accuracy:
60.0 %

Accuracy Comparison
Decision Tree: 56.67 %
KNN: 60.0 %


#**Part F: Linear Regression**


Q19.
Train a Linear Regression model to predict monthly spending.

Q20. Predict the monthly spending for a new user and interpret the result.




In [32]:
# Q-19
X_reg = df.drop('MonthlySpend', axis=1)
y_reg = df['MonthlySpend']

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

lr = LinearRegression()
lr.fit(X_train_reg, y_train_reg)

# Q-20
new_user = pd.DataFrame({
    'UserID': [2000],
    'Age': [30],
    'Gender': [1],
    'SubscriptionType': [2],
    'WatchHoursPerWeek': [15],
    'DevicesUsed': [3],
    'FavoriteGenre': [1],
    'AdClicks': [20],
    'SubscriptionRenewed': [1]
})

prediction = lr.predict(new_user)

print("\nPredicted Monthly Spending:")
print("₹", round(prediction[0], 2))



Predicted Monthly Spending:
₹ 1310.4


#**Business Reflection Questions**


---



**1. Which factors appear to influence subscription renewal the most?**

Ans-
- Based on the Decision Tree feature importance:
    * MonthlySpend
    * UserID (not a meaningful business feature)
    * Age
    * WatchHoursPerWeek

**2. Why is subscription renewal a classification problem?**

Ans-
- Because the target has discrete categories:

Yes

No

The model predicts a class label.

**3. Why is monthly spending a regression problem?**

Ans-
- Monthly spending is a continuous numeric value (₹), so we predict a number rather than a category.

**4. Which algorithm performed better for renewal prediction?**

Ans-
- KNN (60.00%) performed better than Decision Tree (56.67%).

**5. How could the platform use these predictions to improve customer retention?**

Ans-
- Identify users likely to cancel.
- Offer personalized discounts.
- Recommend relevant content.
- Increase customer engagement.
- Improve retention and revenue.